### PR TITLE
Add Python versions and bump ci deps

### PR DESCRIPTION
**Pull request by krisfremen**
## Pull Request Checklist

Thank you for taking the time to improve Arrow! Before submitting your pull request, please check all *appropriate* boxes:

- [ ] 🧪  Added **tests** for changed code.
- [x] 🛠️  All tests **pass** when run locally (run `tox` or `make test` to find out!).
- [x] 🧹  All linting checks **pass** when run locally (run `tox -e lint` or `make lint` to find out!).
- [x] 📚  Updated **documentation** for changed code.
- [x] ⏩  Code is **up-to-date** with the `master` branch.

If you have *any* questions about your code changes or any of the points above, please submit your questions along with the pull request and we will try our best to help!

### CODE REVIEW

**Review by jadchaar — APPROVED** [review]
LGTM

File: .github/workflows/continuous_integration.yml [hunk 1]
```diff
@@ -47,27 +47,27 @@ jobs:
...
           file: coverage.xml
 
   linting:
     name: Linting
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v3
-      - uses: actions/cache@v3
+      - uses: actions/checkout@v4
+      - uses: actions/cache@v4
```
Comment (jadchaar) [inline .github/workflows/continuous_integration.yml:59]: Someone recently added dependabot for these. We should be able to rely on that for github actions, but the rest of the version bumps look good. 

Mind removing the GH action updates so we can let Dependabot take care of it?

Comment (krisfremen) [reply .github/workflows/continuous_integration.yml]: done

File: .github/workflows/continuous_integration.yml [hunk 2]
```diff
@@ -15,7 +15,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: ["pypy-3.9", "3.8", "3.9", "3.10", "3.11", "3.12-dev"]
+        python-version: ["pypy-3.9", "3.8", "3.9", "3.10", "3.11", "3.12", "3.13-dev"]
```
Comment (jadchaar) [inline .github/workflows/continuous_integration.yml:18]: Isn't 3.13 released now? Any reason to be using dev anymore?

**Review by jadchaar — APPROVED** [review]

File: .github/workflows/continuous_integration.yml (hunk 2 above)
Comment (krisfremen) [reply .github/workflows/continuous_integration.yml]: as of 2 weeks ago, I guess that's the case, yeah. I'll drop dev.

**Review by jadchaar — APPROVED** [review]

**Pull request merged**